# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: xpfe/appshell/nsAppShellService.cpp
# Commit: 6fd57b83bb52
# Full Hash: 6fd57b83bb52e9683bade8cbfb5f8ba6575ac6af
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-02-07 03:55:51
# Regressor Bug: 1582832
# File Overlap Count: 3
# Description:
#   Bug 1582832: Part 1 - Make FrameLoader owner rather than DocShell responsible for discarding a BC. r=nika
#   
#   There are all sorts of lifecycle issues which arise from making DocShell
#   responsible for discarding BrowsingContexts. In this particular bug, we tend
#   to run into them in cases where we create a BrowsingContext for a FrameLoader,
# ==============================================================================

diff -r 28a9e4b075af -r 6fd57b83bb52 xpfe/appshell/nsAppShellService.cpp
--- a/xpfe/appshell/nsAppShellService.cpp	Thu Feb 06 18:35:10 2020 +0000
+++ b/xpfe/appshell/nsAppShellService.cpp	Thu Feb 06 19:07:56 2020 +0000
@@ -321,11 +321,25 @@
         mBrowser(aBrowser),
         mContainer(aContainer) {}
 
+  static nsresult Destroy(nsIWebBrowser* aBrowser) {
+    RefPtr<BrowsingContext> bc;
+    if (nsCOMPtr<nsIDocShell> docShell = do_GetInterface(aBrowser)) {
+      bc = docShell->GetBrowsingContext();
+    }
+
+    nsCOMPtr<nsIBaseWindow> window(do_QueryInterface(aBrowser));
+    nsresult rv = window->Destroy();
+    MOZ_ASSERT(bc);
+    if (bc) {
+      bc->Detach();
+    }
+    return rv;
+  }
+
   NS_IMETHOD
   Run() override {
     // Explicitly destroy the browser, in case this isn't the last reference.
-    nsCOMPtr<nsIBaseWindow> window = do_QueryInterface(mBrowser);
-    return window->Destroy();
+    return Destroy(mBrowser);
   }
 
  protected:
@@ -354,8 +368,8 @@
   NS_FORWARD_SAFE_NSIWEBNAVIGATION(mWebNavigation)
   NS_FORWARD_SAFE_NSIINTERFACEREQUESTOR(mInterfaceRequestor)
 
- protected:
-  virtual ~WindowlessBrowser() {
+ private:
+  ~WindowlessBrowser() {
     if (mClosed) {
       return;
     }
@@ -366,11 +380,10 @@
     // when it's safe to run scripts. If this was triggered by GC, it may
     // not always be safe to run scripts, in which cases we need to delay
     // destruction until it is.
-    nsCOMPtr<nsIRunnable> runnable = new BrowserDestroyer(mBrowser, mContainer);
-    nsContentUtils::AddScriptRunner(runnable);
+    auto runnable = MakeRefPtr<BrowserDestroyer>(mBrowser, mContainer);
+    nsContentUtils::AddScriptRunner(runnable.forget());
   }
 
- private:
   nsCOMPtr<nsIWebBrowser> mBrowser;
   nsCOMPtr<nsIWebNavigation> mWebNavigation;
   nsCOMPtr<nsIInterfaceRequestor> mInterfaceRequestor;
@@ -393,9 +406,7 @@
 
   mWebNavigation = nullptr;
   mInterfaceRequestor = nullptr;
-
-  nsCOMPtr<nsIBaseWindow> window = do_QueryInterface(mBrowser);
-  return window->Destroy();
+  return BrowserDestroyer::Destroy(mBrowser);
 }
 
 NS_IMETHODIMP
